### PR TITLE
wr-42: files uploading

### LINKED ISSUE #42: Files upload functionality
- [x] create files table
- [x] implement file saving flow

  - files model
  - files repository
  - files service: should provide capabilities to put file object on [AWS S3](https://aws.amazon.com/ru/s3/)
  - files controller: should provide route for file upload
  - files plugin: should provide custom fastify plugin for files upload validation (check file extension: `jpeg`/`png`/`webp`/`bmp`/`svg+xml`,, check file size: `< 10Mb`)

### PR DESCRIPTION
**Pull request by artemmatiushenko1**
Implemented files package (controller, service, model, entity, repository) to allow API users to upload files. Implemented file type and size validation.

### CODE REVIEW

File: backend/src/libs/plugins/file-upload/libs/types/file-to-upload.type.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+type FileToUpload = {
+  buffer: Buffer;
+  extenstion: string;
```
Comment (andanf-e) [inline backend/src/libs/plugins/file-upload/libs/types/file-to-upload.type.ts]:
```suggestion
  extension: string;
```

File: backend/src/packages/files/file.controller.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  private async upload({
+    fileToUpload,
+  }: ApiHandlerOptions<{
+    fileToUpload: FileToUpload;
+  }>): Promise<ApiHandlerResponse> {
+    return {
+      status: HttpCode.OK,
+      payload: await this.fileService.upload(
+        fileToUpload.buffer,
+        fileToUpload.extenstion,
```
Comment (andanf-e) [inline backend/src/packages/files/file.controller.ts]:
```suggestion
        fileToUpload.extension,
```

File: backend/src/libs/plugins/file-upload/file-upload.plugin.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+        if (data.file.truncated) {
+          throw new FileSizeLimitExceededError();
+        }
+
+        const fileBuffer = await data.toBuffer();
+
+        request.fileToUpload = {
+          buffer: fileBuffer,
+          extenstion: getFileExtensionFromFileName(data.filename),
```
Comment (andanf-e) [inline backend/src/libs/plugins/file-upload/file-upload.plugin.ts]:
```suggestion
          extension: getFileExtensionFromFileName(data.filename),
```

File: backend/src/libs/packages/file/libs/helpers/helpers.ts
Comment (andanf-e) [inline backend/src/libs/packages/file/libs/helpers/helpers.ts:1]: Typo also in file name

File: backend/src/libs/packages/file/libs/helpers/get-file-extenstion-from-filename.helper.ts
Comment (andanf-e) [inline backend/src/libs/packages/file/libs/helpers/get-file-extenstion-from-filename.helper.ts:1]: Type in file name `extension`

File: backend/src/packages/files/file-upload-client.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  private getFileUrl(key: string): string {
+    const { AWS } = this.config.ENV;
+
+    return `https://${AWS.AWS_BUCKET_NAME}.s3.${AWS.AWS_REGION}.amazonaws.com/${key}`;
+  }
+
+  private generateFileKey(extension: string): string {
+    return 'file_' + Date.now().toString() + extension;
```
Comment (andanf-e) [inline backend/src/packages/files/file-upload-client.ts]: Use template string for concat

**Review by andanf-e — CHANGES_REQUESTED** [review]

**Review by andanf-e — DISMISSED** [review]

**Comment by artemmatiushenko1** [comment]
@andanf-e I've managed to move declaration merging to index.d.ts. I not sure how, but it currently works. Maybe it was a bug specifically on my machine

File: backend/src/libs/packages/file/libs/constants/constants.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import { type ValueOf } from 'shared/build/index.js';
```
Comment (andanf-e) [inline backend/src/libs/packages/file/libs/constants/constants.ts]:
```suggestion
import { type ValueOf } from '~/libs/types/types.js';
```

File: backend/src/libs/packages/file/libs/helpers/mb-to-bytes.helper.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
+const mbToBytes = (mb: number): number => {
```
Comment (andanf-e) [inline backend/src/libs/packages/file/libs/helpers/mb-to-bytes.helper.ts]:
```suggestion
const convertMbToBytes = (mb: number): number => {
```

File: backend/src/index.d.ts [hunk 7]
```diff
@@ -1,7 +1,12 @@
+import {
+  type FileToUpload,
+  type FileUploadDecoratedPropety,
+} from './libs/plugins/file-upload/file-upload.js';
 import { type UserAuthResponseDto } from './packages/users/users.js';
 
 declare module 'fastify' {
   interface FastifyRequest {
     user: UserAuthResponseDto;
+    [FileUploadDecoratedPropety.FILE_TO_UPLOAD]: FileToUpload | null;
```
Comment (andanf-e) [inline backend/src/index.d.ts]: Typo
```suggestion
    [FileUploadDecoratedProperty.FILE_TO_UPLOAD]: FileToUpload | null;
```

**Review by andanf-e — CHANGES_REQUESTED** [review]

**Review by NazarKovtun1 — DISMISSED** [review]

**Pull request merged**